mountpoint-s3/src/fs.rs

```diff
@@ -764,19 +769,77 @@ where
         };
 
         if offset != dir_handle.offset() {
+            // POSIX allows seeking an open directory. That's a pain for us since we are streaming
+            // the directory entries and don't want to keep them all in memory. But one common case
+            // we've seen (https://github.com/awslabs/mountpoint-s3/issues/477) is applications that
+            // request offset 0 twice in a row. So we remember the last response and, if repeated,
+            // we return it again.
+            let last_response = dir_handle.last_response.lock().await;
+            if let Some((last_offset, entries)) = &*last_response {
+                if offset == *last_offset {
+                    trace!(offset, "repeating readdir response");
+                    for entry in entries {
+                        if reply.add(entry.clone()) {
+                            break;
+                        }
```

Comment:
I'm a bit uncomfortable with this right now.

If the buffer is full, we should be keeping the entries we've not returned and return them in the next call? I'm guessing this typically works anyway since the entries are the same and the buffer presumably will be the same size, but I'd prefer we handled this case with an error and log that this case occurred.

Reply:
I thought about doing this, but it seemed a bit too much complexity -- we'd have to pop the used entries off the vector and update the offset or something. The contract for readdir is just "return as many entries as you can fit into the buffer, starting from this offset", which this satisfies. If the buffer got smaller (which won't happen in practice), it seems fine to just fulfil the contract the first time around and give up on the second attempt.

Reply:
Ah nice! I hadn't noticed that we can still continue so long as the next offset is what we expect from the first request. Yep, agree to just error out if its one of those weird edge cases.

Thanks for the new tests, that's clear now.